### PR TITLE
Implement restart functionality

### PR DESCRIPTION
**Pull request by dgw**
This PR is an up-to-date version of the seemingly abandoned #1103 by @pachuco.

Personally, I need to dig more into why the `pid_file` parameter exists in `run()`, and why it was removed in this PR. But any further changes I might make to this pull request as a result will be in new commits, so as not to overwrite @pachuco's author credits in the commit log.

Assigning to 7.0.0 for now, to match the tentative (but never official) placement of the original PR. (Target milestone subject to change, of course.)

### CODE REVIEW

**Comment by dgw** [comment]
Best I can tell, `pid_file` was added to `run()` for Reasons™ way back. As best I can tell from testing locally, it looks like the PID file is handled just fine with these changes, so I'll strike that note from the OP.

After looking into this, though, I think a `--restart` command-line option to go with `--quit` and `--kill` would be appropriate. I'll get to that (and am self-assigning the PR as a reminder), so for now I'll remove the "Needs Review" label because the PR's not really finished any more.

**Comment by dgw** [comment]
Still WIP on the CLI front, but I did notice that `bot.restart()` didn't actually work and fixed that separately.

**Comment by dgw** [comment]
Rebased on `master` to resolve merge conflict.

**Comment by Exirel** [comment]
I'm so sorry for the merge conflict all over the place. :D

File: sopel/run_script.py [hunk 1]
```diff
@@ -330,7 +343,14 @@ def main(argv=None):
             pid_file.write(str(os.getpid()))
 
         # Step Seven: Initialize and run Sopel
-        run(config_module, pid_file_path)
+        ret = run(config_module, pid_file_path)
+        os.unlink(pid_file_path)
+        if ret == -1:
+            print(sys.argv)
```
Comment (Exirel) [inline sopel/run_script.py]: Is it for debug purpose?

Comment (dgw) [reply sopel/run_script.py]: Probably. I'll remove it.

**Pull request merged**